Model and LLVM IR to MLIR conversion

vasama-work opened the pull request.

pfez commented on lib/mlir/Dialect/Clift/IR/CliftAttributes.cpp:
```diff
@@ -35,7 +35,7 @@ void mlir::clift::CliftDialect::registerAttributes() {
 mlir::LogicalResult
 mlir::clift::FieldAttr::verify(llvm::function_ref<mlir::InFlightDiagnostic()>
                                  EmitError,
-                               unsigned long Offset,
```
Why not `size_t` instead of `uint64_t`?

pfez replied:
Same below in this commit.

vasama-work replied:
`size_t` is a host type. It could be 32 bits. `uint64_t` is what the model uses.

pfez replied:
Makes sense

pfez commented on lib/mlir/Dialect/Clift/IR/CliftAttributes.cpp:
```diff
@@ -209,21 +221,17 @@ AttrType parseImpl(mlir::AsmParser &parser, llvm::StringRef TypeName) {
-      Iterator != CurrentlyPrintedTypes.end()) {
+  if (const auto Attr = getVisitedType(ID)) {
     if (parser.parseGreater().failed()) {
-      return OnUnepxectedToken(">");
```
This should go in the other commit where you fix the spelling.

All the commits should compile.

pfez commented on lib/mlir/Dialect/Clift/IR/CliftTypes.cpp:
```diff
@@ -179,73 +179,78 @@ using namespace mlir::clift;
-      }
-      lastValue = field.getRawValue();
-    }
+  const auto PrimitiveType = UnderlyingType.cast<mlir::clift::PrimitiveType>();
```
This crashes, while before we had a helpful error message. I think `isa` followed by `cast` is still superior.

vasama-work replied:
The `isa` is still there, on line 185. I just renamed `element_type` to `UnderlyingType`. We do prefer `PascalCase` for locals, right?

pfez replied:
Ah sorry, the side-by-side diff confused me because it was much higher in the page.

We do prefer `PascalCase`.
And BTW, thanks for fixing the casing. It was inconsistent.

pfez commented on lib/mlir/Dialect/Clift/IR/CliftTypes.cpp:
```diff
@@ -179,73 +179,78 @@ using namespace mlir::clift;
-      lastValue = field.getRawValue();
-    }
+  const auto PrimitiveType = UnderlyingType.cast<mlir::clift::PrimitiveType>();
+  const uint64_t Width = PrimitiveType.getSize() * 8;
```
`BitWidth` maybe? It's more in line with the usual LLVM terminology.

vasama-work replied:
I'll change that.

pfez commented on tests/unit/CliftTypes.cpp:
```diff
@@ -0,0 +1,51 @@
+#include "revng-c/mlir/Dialect/Clift/IR/Clift.h"
+#include "revng-c/mlir/Dialect/Clift/IR/ImportModel.h"
+
+#include "ModelTypes.inc"
```
What's with the `.inc` filename? I don't thing we ever use them anywhere else. Just make it a regular `.h` please.

vasama-work replied:
See here: https://rev.ng/gitlab/revng-private/revng-c/-/merge_requests/193#note_23454

pfez replied:
Thanks for pointing to the relevant conversation. It makes sense.

pfez commented on include/revng-c/mlir/Dialect/Clift/IR/CliftAttributes.h:
```diff
@@ -42,40 +42,40 @@ class StructType
-             uint64_t ID);
+             MLIRContext *Context,
+             uint64_t ID,
+             llvm::StringRef Name,
```
:thinking: why do we need the name now?

Also, even if we really need it, being it optional I'd leave it as last argument (which I think also leads to be able to drop this commit ae216398ab4788bc44567adb8febb1b76916be0c )

pfez replied:
Same for the previous `get` method.

vasama-work replied:
We don't *need* the name yet. There's two ways to look at the split:
Should we take up front only the things we really need? Or should we take up front everything that we can and only specify later the things that can only be specified later? I went with the latter, but I think either is fine.
I made it the last argument mainly for symmetry with `UnionType`. This way they have a common prefix. This was mainly important for the storage type template, but I kept it consistent for the attribute types.

pfez replied:
I think we should take upfront only what we really need.
Do you have any compelling reason to do the opposite?

If we do like I propose, AFAICT then both `UnionType` and `StructType` have a common prefix, that is needed by both, and it's only the ID. Makes sense?

vasama-work replied:
I changed this in the `Review changes 2 (to be squashed)` commit. Once you're happy with the changes, I'll squash it into the previous commits.

pfez replied:
Thanks!

pfez commented on include/revng-c/mlir/Dialect/Clift/IR/CliftAttributes.h:
```diff
@@ -122,34 +124,36 @@ class UnionType : public Attribute::AttrBase<UnionType,
   using Base::Base;
 
-  static UnionType get(MLIRContext *ctx, uint64_t ID);
+  static UnionType get(MLIRContext *Context, uint64_t ID, llvm::StringRef Name);
```
Ok good, here it's the last!

pfez commented on include/revng-c/mlir/Dialect/Clift/IR/CliftAttributes.h:
```diff
@@ -42,40 +42,40 @@ class StructType
+             MLIRContext *Context,
+             uint64_t ID,
+             llvm::StringRef Name,
+             uint64_t Size);
```
Also, why do we need the `Size`? I think that the `Size` should always be known, not something that is attached afterwards when the struct is completed.

vasama-work replied:
That is correct. This is the function which does not take the field array and produces a potentially incomplete attribute.

pfez commented on include/revng-c/mlir/Dialect/Clift/IR/CliftStorage.h:
```diff
@@ -12,166 +14,109 @@
+  struct KeyBase {
     uint64_t ID;
-    llvm::StringRef name;
-    uint64_t Size;
```
Ok so the `Size` cannot be part of the `StructTypeStorage` now? How so? It must have to do with the fact that you allow it to be added afterwards when you complete the type. Is that the case?

pfez replied:
Or is it just to allow to CRTP all the `*TypeStorage` stuff?

vasama-work replied:
`StructTypeStorage` and `UnionTypeStorage` are almost the same, the only difference being the additional `uint64_t Size` member in the `StructTypeStorage` key type. That's why I factored out the common parts into `ClassTypeStorage`.
`uint64_t Size` is still there: https://github.com/revng/revng-c/pull/1/files#diff-fa79273e9b99570adf298a58172247aa90dfa23f27e2c87a9b94c7fa0ff8c9d4R107

pfez replied:
:+1:

pfez commented on include/revng-c/mlir/Dialect/Clift/IR/CliftStorage.h:
```diff
@@ -12,166 +14,109 @@
-    if (Key.isInitialized()) {
-      auto Res = ToReturn->mutate(allocator, Key.name, Key.Size, *Key.fields);
-      revng_assert(Res.succeeded());
+    friend bool operator==(const KeyBase &LHS, const KeyBase &RHS) {
```
Why `friend` now?

pfez replied:
Ah, because of CRTP I guess?

vasama-work replied:
There's not a big difference between member operators and hidden friends.

```CPP
friend bool operator==(T const& lhs, T const& rhs)
{
  return lhs.something == rhs.something;
}
// vs
bool operator==(T const& other)
{
  return this->something == other.something;
}
```

Hidden friends are more symmetrical and behave better in some cases where there are implicit conversions in play. Generally hidden friends are recommended over member operators.

To clarify, a hidden friend is a friend function defined within a class definition. It is called *hidden*, because it only participates in overload resolution through ADL.

pfez commented on lib/mlir/Dialect/Clift/IR/CliftAttributes.cpp:
```diff
@@ -282,30 +288,30 @@ AttrType parseImpl(mlir::AsmParser &parser, llvm::StringRef TypeName) {
-    Attrs = ParserType::parse(parser);
-  if (::mlir::failed(Attrs)) {
-    parser.emitError(parser.getCurrentLocation(),
-                     "failed to parse Clift_EnumAttr parameter 'fields' "
```
`Clift_EnumAttr`? I thought it was parsing struct/union here...

pfez commented on include/revng-c/mlir/Dialect/Clift/IR/CliftStorage.h:
```diff
@@ -12,166 +14,109 @@
-    }
+  template<typename... ArgsT>
+  ClassTypeStorage(const uint64_t ID,
+                   const llvm::StringRef Name,
```
Just FYI, `llvm::StringRef` already has the semantic of an immutable reference, so `const` is not really strictly necessary.

vasama-work replied:
`const` here applies to the `StringRef` object, preventing its reassignment, and not to the `char`s to which it refers.
`const int* p;` vs `int* const p;`
This is really just a personal preference thing. I always mark as const any locals that are not mutated. Improves correctness and readability.

pfez replied:
Yeah sure. I have the same preference. I just took the opportunity to note this fact about `StringRef` being an immutable reference, which kinda surprised me the first time I realized it.

pfez commented:
Overall, from a preliminary look, this PR seems to be going in a good direction!

I have some batched comments that I will send soon after I finish the real review.
Meanwhile, I have a broader comment on all the patchset, that I couldn't attach to any specific file, and for which I'd like to involve @aleclearmind on.

In many places you remove checks from `verify` that enforce uniqueness of names.
Now, I know that in the design document we say all over the place that names are just eye-pleasing stuff mainly useful for debug, but are we really sure that we want to allow non-empty names to be duplicated in contexts where they would clash?
It could mean that you're trying to debug something and you have two fields in a struct with the same name, and at some point you have a field access and you have no way of knowing which of the two is being accessed.

pfez commented on tests/unit/mlir_lit_tests/model-import/RawFunctionType.yml:
```diff
@@ -0,0 +1,3 @@
+#
+# Copyright rev.ng Labs Srl. See LICENSE.md for details.
+#
```
Reminder to add this after RFT is stabilized.

pfez commented on include/revng-c/mlir/Dialect/Clift/IR/ImportModel.h:
```diff
@@ -0,0 +1,36 @@
+getQualifiedTypeChecked(llvm::function_ref<mlir::InFlightDiagnostic()>
+                          EmitError,
+                        mlir::MLIRContext &Context,
+                        const model::QualifiedType &ModelType);
```
Please add doxygen documentation to these functions.

In particular, what's the difference between the `Checked` and the non-`Checked` versions? What happens on failure of the check for the `Checked` versions?

vasama-work replied:
I added some documentation and got rid of the variants. Now it's just two `importModelType` overloads for `Type` and `QualifiedType`.

pfez commented on lib/mlir/Dialect/Clift/IR/ImportModel.cpp:
```diff
@@ -0,0 +1,599 @@
+  }
+
+  static clift::PrimitiveKind
+  getPrimitiveKind(const model::PrimitiveTypeKind::Values K) {
```
`toCliftPrimitiveKind` maybe?

Also, taking a `const model::PrimitiveType &` and switching on its kind makes the callers less verbose.

vasama-work replied:
I'll change it to take `const PrimitiveType &` but keep the name if that's alright.

pfez commented on lib/mlir/Dialect/Clift/IR/ImportModel.cpp:
```diff
@@ -0,0 +1,599 @@
+    case model::PrimitiveTypeKind::Float:
+      return clift::PrimitiveKind::FloatKind;
+    default:
+      revng_abort();
```
`revng_abort` can also take a string literal used as error message. Add something meaningful here.

vasama-work replied:
I fixed this and added an aborting case for the `Invalid` and `Count` fields instead of using the default case. This way if any new fields are added in the future we get an error here.

pfez commented on lib/mlir/Dialect/Clift/IR/ImportModel.cpp:
```diff
@@ -0,0 +1,599 @@
+
+  llvm::SmallSet<uint64_t, 16> DefinitionGuardSet;
+
+  class DefinitionGuard {
```
`RecursiveDefinitionGuard` maybe? More verbose but more meaningful.

pfez commented on lib/mlir/Dialect/Clift/IR/ImportModel.cpp:
```diff
@@ -0,0 +1,599 @@
+  }
+
+  RecursiveCoroutine<clift::ValueType>
+  getQualifiedType(const model::QualifiedType &ModelType,
```
`getValue` type I guess. Otherwise a caller expects to get a `QualifiedType`, not a `clift::ValueType`.

vasama-work replied:
I renamed it to `getQualifiedValueType`.

pfez commented on lib/mlir/Dialect/Clift/IR/ImportModel.cpp:
```diff
@@ -0,0 +1,599 @@
+      }
+    }
+
+    clift::ValueType QualifiedType = rc_recur
```
Using `QualifiedType` as variable name is confusing.
`UnwrappedType` maybe?

vasama-work replied:
I renamed it to `ResultType`.

pfez commented on lib/mlir/Dialect/Clift/IR/ImportModel.cpp:
```diff
@@ -0,0 +1,599 @@
+
+    // If the set of qualifiers contains any pointers,
+    // the base type does not need to be complete.
+    for (const model::Qualifier &Qualifier : Qualifiers) {
+      if (Qualifier.Kind() == model::QualifierKind::Pointer) {
+        RequireComplete = false;
+        break;
+      }
+    }
```
```c++
if (llvm::any_of(Qualifiers, model::Qualifier::isPointer))
  RequireComplete = false;
```

pfez commented on lib/mlir/Dialect/Clift/IR/ImportModel.cpp:
```diff
@@ -0,0 +1,599 @@
+
+    const auto takeConst = [&]() -> bool {
+      if (not Qualifiers.empty()
+          and Qualifiers.back().Kind() == model::QualifierKind::Const) {
```
```c++
model::Qualifer::isConst(Qualifiers.back())
```

pfez commented on lib/mlir/Dialect/Clift/IR/ImportModel.cpp:
```diff
@@ -0,0 +1,599 @@
+                                     Fields);
+  }
+
+  RecursiveCoroutine<clift::TypeDefinition>
+  getTypeAttribute(const model::Type &ModelType, bool &RequireComplete) {
+    switch (ModelType.Kind()) {
+    case model::TypeKind::CABIFunctionType: {
+      RequireComplete = true;
+      auto const &T = llvm::cast<model::CABIFunctionType>(ModelType);
+      return getTypeAttribute(T);
+    }
+    case model::TypeKind::EnumType: {
+      RequireComplete = true;
+      auto const &T = llvm::cast<model::EnumType>(ModelType);
+      return getTypeAttribute(T);
+    }
+    case model::TypeKind::RawFunctionType: {
+      RequireComplete = true;
+      auto const &T = llvm::cast<model::RawFunctionType>(ModelType);
+      return getTypeAttribute(T);
+    }
+    case model::TypeKind::StructType: {
+      auto const &T = llvm::cast<model::StructType>(ModelType);
+      return getTypeAttribute(T, RequireComplete);
+    }
+    case model::TypeKind::TypedefType: {
+      auto const &T = llvm::cast<model::TypedefType>(ModelType);
+      return getTypeAttribute(T, RequireComplete);
+    }
+    case model::TypeKind::UnionType: {
+      auto const &T = llvm::cast<model::UnionType>(ModelType);
+      return getTypeAttribute(T, RequireComplete);
+    }
+    default:
+      revng_abort();
+    }
+  }
```
The function returns a `RecursiveCoroutine`, so all the `return`s should be `rc_return rc_recur getTypeAttribute`.

I'm honestly surprised this even compiles. My expectation was that the compiler should warn you or plainly error out :thinking:

vasama-work replied:
`RecursiveCoroutine` is just a regular object and returning one from a function is just fine. In a coroutine this object is constructed implicitly by the compiler. We're just returning the object returned by another coroutine. Think of this as a coroutine tail call. It's just more efficient this way.

pfez replied:
Fair enough. It's just not something I had in mind when I designed them, but it checks out!

Finally someone is using them besides me, and this brings up new ways to use them that I did not envision :)

pfez commented on lib/mlir/Dialect/Clift/IR/ImportModel.cpp:
```diff
@@ -0,0 +1,599 @@
+
+  mlir::BoolAttr getFalse() { return getBool(false); }
+
+  template<typename T, typename... ArgTypes>
```
Can't we use a `concept` to put some constraints on `T`.
The only valid `T`s are `clift::ValueType`s right?

vasama-work replied:
It's any MLIR type with static `verify` and `get` functions. It's used for both types and attributes. Maybe I could constrain it somehow, but that doesn't seem particularly useful since it's a very local function.

pfez commented on lib/mlir/Dialect/Clift/IR/ImportModel.cpp:
```diff
@@ -0,0 +1,599 @@
+      rc_return nullptr;
+
+    rc_return make<clift::FunctionAttr>(ModelType.ID(),
+                                        ModelType.OriginalName(),
```
I'm putting this comment here, but it applies to any other place dealing with `OriginalName` vs `CustomName`.

Basically, this is linked to the issue of uniqueness of names.
`CustomName`s are already guaranteed to be uniqued by the Model where relevant.
So one thing we could do is: switch back to requiring uniqueness, and use `CustomName()` everywhere instead of `OriginalName()`, to get uniqueness from the model for free. If anybody ever writes another frontend, they'll have to deal with uniqueness in some other way, but for us for now it's not a problem.
It seems to me like a path of low impedance, not necessarily the best design in the long term.

In any case, whatever we decide about uniqueness of names, the clift type importer should be self-consistent and either always use `OriginalName()` or `CustomName()` everywhere. Now some places use `OriginalName`, some others `CustomName`. Any specific reason for that?

vasama-work replied:
> Any specific reason for that?

No, I don't think so. Let's pick one and I'll change them to be consistent.

vasama-work replied:
After more discussion on Mattermost, these were all changed to use `name()` and `entryName()` for enum fields.

pfez commented on lib/mlir/Dialect/Clift/IR/ImportModel.cpp:
```diff
@@ -0,0 +1,599 @@
+  }
+
+  RecursiveCoroutine<clift::TypeDefinition>
+  getTypeAttribute(const model::RawFunctionType &ModelType) {
```
Adding this comment to remind me that I have to re-review this function after the latest updates to RFT that are still ongoing.

pfez commented on tests/unit/ModelTypes.inc:
```diff
@@ -0,0 +1,631 @@
+    }
+  }
+
+  revng_check(not PrimitiveType::fromName("foalt32_t"));
+  revng_check(not PrimitiveType::fromName("generic7_t"));
+  revng_check(not PrimitiveType::fromName("generic8"));
```
:brain:

pfez commented on tests/unit/ModelTypes.inc:
```diff
@@ -0,0 +1,631 @@
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
```
Don't know where to put this, but really thanks a lot for this effort of testing stuff.

vasama-work replied:
Unfortunately I can't take credit for all this test code. I just adapted the model type test for Clift.

pfez commented on tests/unit/CliftTypes.cpp:
```diff
@@ -0,0 +1,51 @@
+  return Tree.verify(false);
+}
+
+static bool checkSerialization(const TupleTree<model::Binary> &Tree) {
+  return true;
+}
```
:thinking: I'm puzzled by this. What's the purpose of having a test that always returns `true`? Is this something you plan to finalize in a second moment?

vasama-work replied:
This is for sharing the test code between the model and Clift. This code is taken from here: https://github.com/revng/revng/blob/develop/tests/unit/ModelType.cpp
These static functions are "customisations" of the test where different functions are provided in the model and Clift tests. There's no need to test model serialisation again in the Clift test.

pfez commented on tests/unit/ModelTypes.inc:
```diff
@@ -0,0 +1,631 @@
+  revng_check(verify(*T, true));
+  revng_check(checkSerialization(T));
+
+  // Assigning succeeds if even if an index is already present
```
if even if

pfez commented on tests/unit/ModelTypes.inc:
```diff
@@ -0,0 +1,631 @@
+  revng_check(checkSerialization(T));
+}
+
+BOOST_AUTO_TEST_CASE(RawFunctionTypes) {
```
Note to self. Re-review this after we finalized the ongoing changes to RFT.

vasama-work committed:
Fix spelling of OnUnexpectedToken

vasama-work committed:
Replace unsigned long with uint64_t

vasama-work committed:
Add missing verify functions

vasama-work committed:
Replace MLIR container and LLVM import pipe

vasama-work committed:
Remove UnionType::verify size parameter

vasama-work committed:
Fix PointerType::verify

Allow pointers with sizes 4 or 8, matching the model.

vasama-work committed:
Factor Clift type parsing out into separate files

vasama-work committed:
Add model types to Clift MLIR import

vasama-work committed:
Fix EnumAttr::verify

* Fix enumerator value strict ordering check.
* Remove enumerator name uniqueness check.

vasama-work committed:
Add LIT/FileCheck tests for model type import

vasama-work committed:
Fix no-op asserts

vasama-work committed:
Add support for typedef as enum underlying type

vasama-work committed:
Fix error message in StructType::parse

vasama-work committed:
Fix StructType::verify

* Simplify offset and size checks.
* Add field type completeness check.
* Remove field name uniqueness check.

vasama-work committed:
Fix UnionType::verify

* Add field type completeness check.
* Remove field name uniqueness check.

vasama-work committed:
Move struct,union get,getChecked defs to .cpp

There seem to be some issues with defining these in headers due to the
storage type being incomplete. The storage type is only defined in the
.cpp file.

vasama-work committed:
Reorder StructTypeStorage fields

Looking at the MLIR code, the order should match the order of the
parameters in get and verify.

aleclearmind commented on include/revng-c/mlir/Pipes/MLIRContainer.h:
```diff
@@ -0,0 +1,65 @@
+public:
+  static const char ID;
+  static constexpr auto Name = "mlir-module";
+  static constexpr auto MIMEType = "text/mlir";
```
1. Is this bytecode format? Currently for LLVM IR we use the text representation, but that's very verbose, I'd like to switch to the bytecode representation there too.
1. Did we make this up or is there some place where this is said to be the official MIME type of MLIR? For LLVM IR we use `text/x.llvm.ir`. If it's bytecode, it shouldn't start with `text/`.

vasama-work replied:
1. Currently text because it's easier to debug.
2. I didn't find anything official. `text/mlir` was already listed here: https://docs.rev.ng/references/mime-types/

aleclearmind replied:
1. I understand, but these choices tend to stick around forever. Let's switch to byte code. I think you can get the textual representation just by doing `mlir-opt -S` or something like that.
2. Unless we find some official source, I'd go with `application/x.mlir.bc` for MLIR byte code and with `text/x.mlir.ir`, just for coherence with what we're using for LLVM IR.

vasama-work replied:
Done.

aleclearmind commented on include/revng-c/mlir/Pipes/MLIRContainer.h:
```diff
@@ -0,0 +1,65 @@
+  static constexpr auto MIMEType = "text/mlir";
+
+private:
+  // NOTE: Mutability of the context is required for cloneFiltered.
```
Rapidly hint at why it's required

aleclearmind commented on include/revng-c/mlir/Pipes/MLIRContainer.h:
```diff
@@ -0,0 +1,65 @@
+  //       The thread safety of this is not clear at the time of writing,
+  //       but it was decided that thread safety of cloneFiltered was not
+  //       necessary at this time.
+  std::unique_ptr<mlir::MLIRContext> Context;
```
Why the choice of `std::unique_ptr` over a regular `mutable` instance?
Do we need to `std::move` it?

vasama-work replied:
Yes, it needs to be moved. Originally it was mutable, hence the comment.

aleclearmind commented on include/revng-c/mlir/Pipes/MLIRContainer.h:
```diff
@@ -0,0 +1,65 @@
+  //       necessary at this time.
+  std::unique_ptr<mlir::MLIRContext> Context;
+  mlir::OwningOpRef<mlir::ModuleOp> Module;
+  llvm::DenseMap<llvm::StringRef, mlir::FunctionOpInterface> Targets;
```
1. Is this a cache?
   By cache I mean something that could be recomputed from `Module`.

   Caches introduce the risk of going out of sync with the underlying thing they are caching. Unless we have a persuasive reason to think this significantly improves performance, I'd drop it.
2. Why do we use a `StringRef` as a key? Can we use a `MetaAddress`?

vasama-work replied:
I ended up getting rid of the cache entirely. Had been thinking about doing that for a while.

aleclearmind commented on include/revng-c/mlir/Pipes/MLIRContainer.h:
```diff
@@ -0,0 +1,65 @@
+  llvm::DenseMap<llvm::StringRef, mlir::FunctionOpInterface> Targets;
+
+public:
+  MLIRContainer(llvm::StringRef Name);
```
1. We usually keep bodies of small method into the definition. They act a bit as docs and enable better inlining.
2. Is the name mandatory? If not, I'd avoid it.

vasama-work replied:
1. It ended up getting another function call in the body. I can still move it to the header if you prefer.
2. All the other containers seemed to follow this pattern with a constructor taking a name parameter. I'm not entirely clear on its purpose.

aleclearmind replied:
1. I'd say let's keep methods that are 2-3 lines long in the header. That's the typical convention. Basically, if it doesn't hinder readability of the header too much, let's keep it in the header.
2. Let's drop it then.

vasama-work replied:
1. Done
2. To clarify, do you mean drop the issue or drop the parameter?
   Looks like the parameter is needed by `ContainerFactoryWithArgs` [here](https://github.com/revng/revng/blob/develop/include/revng/Pipeline/ContainerFactory.h#L45).

aleclearmind replied:
2. Ah! OK, it's the container name! Then leave it.

aleclearmind commented on include/revng-c/mlir/Pipes/MLIRContainer.h:
```diff
@@ -0,0 +1,65 @@
+  mlir::ModuleOp getModule() { return *Module; }
+  mlir::ModuleOp getOrCreateModule();
+
+  void setModule(mlir::OwningOpRef<mlir::ModuleOp> &&NewModule);
```
Can we drop the `getOrCreateModule` and `setModule` and automatically populate `this->Module` in the constructor?

aleclearmind commented on lib/mlir/Pipes/CMakeLists.txt:
```diff
@@ -9,4 +10,5 @@ target_link_libraries(
          MLIRDLTIDialect
-         MLIRLLVMIRToLLVMTranslation)
+         MLIRLLVMIRToLLVMTranslation
+  PRIVATE MLIRCliftDialect MLIRBytecodeReader MLIRBytecodeWriter)
```
What is the benefit of using `PRIVATE` here?
I'm wary of opportunistically adopting previously unused features around. If you think this is beneficial, make a case for it and we can think of adopting it at the level of the whole codebase.

Coherency of the codebase is often better than adopting features which bring some benefit, but just to a part of the codebase.

vasama-work replied:
It's just generally good practice, akin to "include what you use" and conversely "don't include what you don't use". I can change it to public too if you prefer.

aleclearmind replied:
If I understand the use case of `PRIVATE` for linked libraries correctly, it's used when you have some internal library, that downstream users of your CMake don't need to know about and that you're not exporting in the [CMake stuff you export](https://github.com/revng/revng/blob/develop/CMakeLists.txt#L150).

Not marking such dependency as `PRIVATE` might result in downstream users of the exported stuff to try to depend on it, despite not being exported and therefore fail.

This doesn't seem to be the case, so I'd just leave it out.

vasama-work replied:
I'm not sure on the relation of `PRIVATE`/`PUBLIC` dependencies and exported targets. Within a single project both dependencies are linked transitively, but only the include paths of public dependencies are added transitively. That is, a dependency would only need to be public when it is exposed in an interface.

aleclearmind replied:
I see your point but I don't think in practice we're saving any include path. For each LLVM subproject you get `-I$PROJECT_NAME/include `. There isn't a per-library `-I` argument.

In absence of a concrete benefit, I'd say to leave it out so we don't have to ask ourselves every time "wait should this be `PRIVATE`?".

vasama replied:
Done.

vasama-work replied:
Done.

aleclearmind commented on lib/mlir/Pipes/MLIRContainer.cpp:
```diff
@@ -0,0 +1,339 @@
+using OwningModuleRef = mlir::OwningOpRef<mlir::ModuleOp>;
+
+static OwningModuleRef cloneModule(mlir::MLIRContext &Context,
+                                   mlir::ModuleOp Module) {
```
Is this a cross-context clone?
If so, I'd got for

```c++
static OwningModuleRef cloneModuleInto(mlir::ModuleOp SourceModule, mlir::MLIRContext &DestinationContext) {
```

vasama-work replied:
Done.

aleclearmind commented on lib/mlir/Pipes/MLIRContainer.cpp:
```diff
@@ -0,0 +1,339 @@
+}
+
+static llvm::StringRef getFunctionName(const pipeline::Target &Target) {
+  revng_check(&Target.getKind() == &TheKind);
+  revng_check(Target.getPathComponents().size() == 1);
```
We don't have a super-strict policy as to when use `revng_check` instead of `revng_assert`.
I'd use it only for, say, security critical stuff.

However, I'm not really sure we'll actually ever disable assertions, even in production builds.

vasama-work replied:
The line of thought was that these are sort of like program inputs so check is more appropriate than assert, but that might have been due to a lack of understanding on how the pipeline works. I can change them to asserts.

aleclearmind replied:
Ah, I see. The only thing we consider direct user input are 1) the model, 2) the input binaries and 3) debug info (and whatnot).

However, we should never abort on user input errors, in fact, the input is always "correct". We should either proceed "as best as we can" or gracefully return an error, never abort. Aborts report a programming error, not an user error.

To be more specific, a pipe should either:
1. "proceed as best as we can", for instance sometimes we emit in the IR a call to an function reporting an unhandled situation (`git grep emitAbort`), or
2.  gracefully fail by performing early checks on the model in `checkPrecondition`. Other than that, a pipe should never abort on user input. If `checkPrecondition` passes, the pipe should *always* terminate successfully.

Then obviously it's not always the case, but those are bugs.

Anyway, the specific checks are not considered user input.

vasama-work replied:
I ended up changing these into asserts.
Correction: The whole function is gone now.

aleclearmind commented on lib/mlir/Pipes/MLIRContainer.cpp:
```diff
@@ -0,0 +1,339 @@
+  revng_check(&Target.getKind() == &TheKind);
+  revng_check(Target.getPathComponents().size() == 1);
+
+  auto &&Name = Target.getPathComponents().front();
```
Make sure you're using `auto` [according to our policies](https://pad.rev.ng/s/SJNUet_ku#).

`auto` is nice, but it makes reading the code (and review it) a bit more difficult.

vasama-work replied:
I ended up getting rid of the whole function along with the target cache.

aleclearmind commented on lib/mlir/Pipes/MLIRContainer.cpp:
Please comment what goes on in these functions. There's several non-obvious thing that deserve a comment to ease reading.

vasama-work replied:
I've added some comments. Hopefully that makes things clearer.

aleclearmind commented on lib/mlir/Pipes/MLIRContainer.cpp:
```diff
@@ -0,0 +1,339 @@
+  return mlir::cast<mlir::FunctionOpInterface>(O).isExternal();
+}
+
+static void pruneSymbols(mlir::ModuleOp Module) {
```
NOTETOSELF: I reviewed `MLIRContainer.h` only, resume reviewing from here.

aleclearmind commented on lib/mlir/Pipes/MLIRPipe.cpp:
I think this change belongs to the next commit.
Remember that all commits should compile.

The following command will check that for you:
```
orc check-branch revng-c
```

vasama-work replied:
The equivalent pipe in `ImportLLVMToMLIRPipe.cpp` replaces this one. The Clift type import pipe in the next commit is new.

aleclearmind replied:
OK, then I got confused. If this commit compiles on its own, mark this as resolved.

vasama-work commented:
> In many places you remove checks from `verify` that enforce uniqueness of names. Now, I know that in the design document we say all over the place that names are just eye-pleasing stuff mainly useful for debug, but are we really sure that we want to allow non-empty names to be duplicated in contexts where they would clash? It could mean that you're trying to debug something and you have two fields in a struct with the same name, and at some point you have a field access and you have no way of knowing which of the two is being accessed.

I've now added those checks back in.

aleclearmind commented on tests/unit/mlir_lit_tests/model-import/RawFunctionType.yml:
```diff
@@ -0,0 +1,3 @@
+#
+# Copyright rev.ng Labs Srl. See LICENSE.md for details.
```
New license notice is:
```
This file is distributed under the MIT License. See LICENSE.md for details.
```

vasama-work committed:
Restrict the dialects available in clift-opt

vasama-work committed:
Add ImportModelType pass, register in clift-opt

aleclearmind commented on lib/mlir/Pipes/MLIRContainer.cpp:
```diff
@@ -0,0 +1,270 @@
+std::unique_ptr<pipeline::ContainerBase>
+MLIRContainer::cloneFiltered(const pipeline::TargetsList &Filter) const {
+  auto NewContainer = std::make_unique<MLIRContainer>(name());
+  if (not getModuleBlock(*Module).empty()) {
```
As a matter of style, in case of `if` statements spanning most of the body of the function, we prefer to negate the condition and have an early return. In case of multiple preliminary checks, it significantly reduces the indentation of the part of the code that matters.

vasama-work replied:
Done.

aleclearmind commented on lib/mlir/Pipes/MLIRContainer.cpp:
```diff
@@ -0,0 +1,270 @@
+}
+
+// Remove any unused non-target symbols.
+static void pruneSymbols(mlir::ModuleOp Module) {
```
`pruneUnusedSymbols`

vasama-work replied:
Done.

aleclearmind commented on lib/mlir/Pipes/MLIRContainer.cpp:
```diff
@@ -0,0 +1,270 @@
+        const auto &SymbolRef = Use.getSymbolRef();
+        revng_assert(SymbolRef.getNestedReferences().empty());
+
+        mlir::Operation *const
```
I understand the desire to use `const` as much as possible, but here it seems to hurt readability with not much in return.

vasama-work replied:
Done.

aleclearmind commented on lib/mlir/Pipes/MLIRContainer.cpp:
```diff
@@ -0,0 +1,270 @@
+    if (F.isExternal())
+      return;
+
+    if (const auto &Uses = mlir::SymbolTable::getSymbolUses(F)) {
```
This piece of code works with symbol tables, but AFAIK we're not going to use symbol tables with clift functions.
Can we make this work also with operations implementing `FunctionOpInterface` but not `Symbol`?

aleclearmind commented on lib/mlir/Pipes/MLIRContainer.cpp:
```diff
@@ -0,0 +1,270 @@
+
+static OwningModuleRef cloneModuleInto(mlir::ModuleOp SourceModule,
+                                       mlir::MLIRContext &DestinationContext) {
+  llvm::SmallString<1024> Buffer;
```
Consider `using namespace mlir` and/or `llvm`, it should increase readability.

vasama-work replied:
I added using declarations for some specific commonly used things but did not add any using directives.

aleclearmind commented on lib/mlir/Pipes/MLIRContainer.cpp:
```diff
@@ -0,0 +1,270 @@
+using ContextPtr = std::unique_ptr<mlir::MLIRContext>;
+using OwningModuleRef = mlir::OwningOpRef<mlir::ModuleOp>;
+
+static OwningModuleRef cloneModuleInto(mlir::ModuleOp SourceModule,
+                                       mlir::MLIRContext &DestinationContext) {
```
Please document this function: why does it serialize the module (to have the source module in a different context)? What's the downside of this approach (pay serialization)? What would be the ideal solution (cross-context cloning)?

The function is rather trivial but the reasons why we do things this way, tradeoffs and alternatives should be documented.

aleclearmind commented on lib/mlir/Pipes/MLIRContainer.cpp:
```diff
@@ -0,0 +1,270 @@
+                               mlir::ParserConfig(&DestinationContext));
+  revng_assert(mlir::succeeded(R));
+
+  auto &Operations = OuterBlock.getOperations();
+  revng_assert(Operations.size() == 1);
+
+  auto DestinationModule = mlir::cast<mlir::ModuleOp>(Operations.front());
+  DestinationModule->remove();
```
Add a comment like "Ah we can't directly deserialize the `ModuleOp`, so we deserialize it into a block, then detach it and return it passing ownership to the caller".

aleclearmind commented on lib/mlir/Pipes/MLIRContainer.cpp:
```diff
@@ -0,0 +1,270 @@
+MLIRContainer::cloneFiltered(const pipeline::TargetsList &Filter) const {
+  auto NewContainer = std::make_unique<MLIRContainer>(name());
+  if (not getModuleBlock(*Module).empty()) {
+    OwningModuleRef NewModule(mlir::cast<mlir::ModuleOp>((*Module)->clone()));
```
Add a brief summary:

```
// 1. We clone in the old context
// 2. We filter out things we're not interested in
// 3. We clone the module into a new context
//
// Not that doing 2) before 3) reduces the cost of serialization/deserialization taking place upon cloning
```
Something like this.

vasama-work replied:
> Not that

Is this a typo? Did you mean "Not that" or "Note that"? In any case, the reason for doing the filtering in the source module is mainly to avoid dragging unused types and attributes into the new context. I'll add an explanation of that.

aleclearmind replied:
It was a "note that", yup.

aleclearmind commented on lib/mlir/Pipes/MLIRContainer.cpp:
```diff
@@ -0,0 +1,270 @@
+
+  mlir::Block &Block = getModuleBlock(*Module);
+  visit(*TemporaryModule, [&](mlir::SymbolOpInterface Symbol) {
+    // Erase an existing symbol with the same name, if one exists.
+    if (auto S = mlir::SymbolTable::lookupSymbolIn(*Module, Symbol.getName()))
+      S->erase();
```
As in the other comment, AFAIK `clift.function`s won't implement the `Symbol` interface.

aleclearmind commented on lib/mlir/Pipes/MLIRContainer.cpp:
```diff
@@ -0,0 +1,270 @@
+  return NewContainer;
+}
+
+void MLIRContainer::mergeBackImpl(MLIRContainer &&Container) {
```
It would be nice if all relevant variables/arguments had a `Source` or `Destination` prefix. Same in `cloneFiltered`.

aleclearmind commented on lib/mlir/Pipes/MLIRLLVMHelpers.h:
```diff
@@ -0,0 +1,64 @@
+
+inline std::optional<llvm::StringRef>
+getFunctionName(mlir::FunctionOpInterface F) {
+  if (const auto Name = getLinkageName(F)) {
```
AFAIU we can have "custom" attributes on operations, at this point there's no point in abusing linkage name.
If you're not up for doing it now, at least leave a big TODO.

vasama-work replied:
I got rid of any uses of the linkage name and added the `revng.function.entry` `StringAttr` attribute containing the `MetaAddress`.

aleclearmind commented on lib/mlir/Pipes/MLIRContainer.cpp:
```diff
@@ -0,0 +1,270 @@
+    if (const auto MaybeTargetName = getFunctionName(F))
+      List.push_back(makeTarget(*MaybeTargetName));
+  });
+  llvm::sort(List);
```
Why? Add a comment.

vasama-work replied:
Done.

aleclearmind commented on lib/mlir/Pipes/MLIRContainer.cpp:
```diff
@@ -0,0 +1,270 @@
+
+  if (not NewModule)
+    return llvm::createStringError(llvm::inconvertibleErrorCode(),
+                                   "Cannot load MLIR/LLVM module.");
```
Why `/LLVM`?

vasama-work replied:
Originally this was just going to be a container for MLIR in the LLVM dialect, because I thought the different dialects would require very different code, but that's no longer the case. I'll fix it.

aleclearmind commented on lib/mlir/Pipes/MLIRContainer.cpp:
```diff
@@ -0,0 +1,270 @@
+  return cloneFiltered(pipeline::TargetsList::List{ Target })->serialize(OS);
+}
+
+std::vector<pipeline::Kind *> MLIRContainer::possibleKinds() {
+  return { &TheKind };
+}
```
Inline in header file.

vasama-work replied:
That requires adding `#include "revng-c/Pipes/Kinds.h"` in the header. Should I do it anyway?

aleclearmind replied:
Yeah

aleclearmind commented on lib/mlir/Pipes/MLIRContainer.cpp:
```diff
@@ -0,0 +1,270 @@
+  return { &TheKind };
+}
+
+std::unique_ptr<mlir::MLIRContext> MLIRContainer::makeContext() {
+  const auto Threading = mlir::MLIRContext::Threading::DISABLED;
+  return std::make_unique<mlir::MLIRContext>(Threading);
+}
```
Demote to `static` function?

vasama-work replied:
Done.

vasama-work committed:
Fix spelling of FunctionEntryMDName

vasama-work committed:
Set revng prefix in MLIR LIT tests

vasama-work committed:
Fix spelling of DefaultRawType

vasama-work committed:
Introduce RawFunctionType::Architecture

vasama-work committed:
Add alias for error emission function_ref

vasama-work committed:
Add import-clift-types pipe

vasama-work committed:
Install MLIRCliftDialect library

aleclearmind commented on lib/mlir/Dialect/Clift/Utils/CMakeLists.txt:
```diff
@@ -0,0 +1,18 @@
+# This file is distributed under the MIT License. See LICENSE.md for details.
+#
+
+add_mlir_dialect_library(
```
Is it necessary to use this? I guess we need this only if we use MLIR tablegen magic.

Let's use our `revng_add_library` function wherever we can.

vasama-work replied:
Done.

aleclearmind commented on tests/unit/mlir_lit_tests/ImportReachableModelTypes.mlir:
This kind of tests are extremely sensible to changes and usually lead to great pain.

1. What is each letter testing? `G`, `F` and `H`.
2. Can you elaborate what we're testing here specifically? This looks like a test that should be end-to-end.

If you think this is a temporary test while we wait for testing the end-to-end dialect conversion, I'm OK, but please leave a comment.

vasama-work replied:
The model contains three functions `g`, `f` and `h`. The test is executed once for each function. Otherwise the test would depend on the order in which the aliases are emitted. I added a comment to that effect.

What we're testing is that the all transitively referenced types, but only those types, are imported. Indirectly it also tests the function metadata cache with the MLIR traits. The test is partially temporary in that it will probably change a little once more Clift import functionality is added. But I expect the only change then will be that we check for `clift.func` instead of `clift.undef`.

aleclearmind replied:
> The model contains three functions g, f and h

Why only two functions are defined in MLIR?

In any case, I really have PTSD from similar tests that hardcode things coming from earlier stages of the pipeline that change frequently. I absolutely do not want that to repeat.

Please minimize as much as you can the test, for instance, drop `revng.function.metadata`. Keep what's strictly barely minimally necessary and then let's look at it again.

In any case, I'd like this test to go away in the next PR, in favor of an end-to-end test, i.e., starting from a binary, lifting it and everything. It would be nice to have tests on intermediate representations but given the unstability of the details of the IR, they are *extremely* painful to maintain.

vasama-work replied:
> Why only two functions are defined in MLIR?

We test that the types relating to the third function which is not present in the MLIR are not imported.

> drop `revng.function.metadata`

That's needed by the function metadata cache.

aleclearmind replied:
Mmh. Let's either drop the test or let's go from source.

Going from source implies:
1. Adding a new source file to `revng-qa`, something [like this](https://github.com/revng/revng-qa/blob/master/share/revng/test/configuration/revng-qa/for-import-types.yml).
2. Adding a new command to `revng-c` (something [like this](https://github.com/revng/revng-c/blob/develop/share/revng/test/configuration/revng-c/detect-stack-size.yml)) that does `revng artifact $MY_STEP --model ${SOURCE}.yml`. If using this approach implies dropping the `clift-opt` pass, I'm OK with that.
   Unfortunately, with `revng artifact` I think you can't request just two functions out of 3, you need to use the lower level tool `revng pipeline`, whose command line is a PITA. But Ivan might help you with that.

Having a test focused on MLIR that depends on implementation details of the serialized CFG has historically been a huge pain that sticks around for years, I don't want it.

vasama-work replied:
Do you mind if we just drop the test for now? It passed and in the next PR we'll do something else anyway.

aleclearmind replied:
I'm OK with that, put a note in your backlog though.

aleclearmind commented on lib/ImportFromC/HeaderToModel.cpp:
```diff
@@ -653,6 +653,8 @@ bool DeclVisitor::VisitFunctionDecl(const clang::FunctionDecl *FD) {
   } else {
     auto TheRetClangType = FD->getReturnType();
     auto TheRawFunctionType = cast<RawFunctionType>(NewType.get());
+    TheRawFunctionType->Architecture() = Model->Architecture();
```
This is not OK, we need an attribute in C to denote the architecture of `RawFunctionType`s. Similar to `_ABI`.

Maybe we can get `_ABI(raw_$ARCHITECTURE)` (e.g., `_ABI(raw_x86_64)`)?

vasama-work replied:
Done.

aleclearmind commented on lib/ImportFromC/HeaderToModel.cpp:
```diff
@@ -893,6 +895,7 @@ bool DeclVisitor::VisitFunctionPrototype(const FunctionProtoType *FP,
     auto DefaultRawType = cast<RawFunctionType>(TheDefaultPrototype.get());
 
     auto FunctionType = cast<RawFunctionType>(NewType.get());
+    FunctionType->Architecture() = Model->Architecture();
```
As a above

aleclearmind commented on include/revng-c/mlir/Dialect/Clift/IR/CliftAttributes.h:
Does this commit ("Remove UnionType::verify size parameter ") compile?

vasama-work replied:
It does. I think that particular function was never used until later in the commit that adds the conversion from model types to Clift types.

aleclearmind commented on lib/mlir/Dialect/Clift/IR/CliftAttributes.cpp:
```diff
@@ -65,30 +65,18 @@ ArgAttr::verify(llvm::function_ref<mlir::InFlightDiagnostic()> EmitError,
+    if (mlir::isa<FunctionAttr>(Type.getElementType()))
       return EmitError() << "function type cannot return another function type";
   }
-  std::set<llvm::StringRef> Names;
```
Can you add a note about why we decided not to do this?

vasama-work replied:
I reverted this commit. These names are not actually used for aliases like some of the others are, but since we're checking the others for uniqueness again, I suppose we might as well check these too.

aleclearmind commented on lib/mlir/Dialect/Clift/IR/CliftTypes.cpp:
```diff
@@ -270,8 +270,8 @@ UnionType::replaceImmediateSubElements(llvm::ArrayRef<mlir::Attribute>
                                          replAttrs,
                                        llvm::ArrayRef<mlir::Type> replTypes)
   const {
-  revng_assert("it does not make any sense to replace the elements of a "
-               "defined Union");
+  revng_abort("it does not make any sense to replace the elements of a "
+              "defined Union");
   return {};
```
Good catch.

Remove `return {}` as well. Elsewhere too.

vasama-work replied:
Done.

aleclearmind commented on lib/mlir/Dialect/Clift/IR/CliftTypes.cpp:
```diff
@@ -175,13 +175,27 @@ ArrayType::verify(::llvm::function_ref<::mlir::InFlightDiagnostic()> emitError,
   return mlir::success();
 }
 
+static mlir::clift::TypedefAttr getTypedefAttr(mlir::Type Type) {
```
Out of curiosity, did you catch this bug while testing using the model type system tests from `revng`?

That'd be proof this wasn't a pointless exercise `:D`.

vasama-work replied:
I forget, but I think it was just something that occurred to me while working on the verifications.

pfez replied:
Actually, the spec says we that the underlying type of an enum can only be a `PrimitiveType`. I don't really think we want to allow typedefs.

aleclearmind commented on lib/mlir/Dialect/Clift/IR/CliftAttributes.cpp:
```diff
@@ -291,12 +294,20 @@ static AttrType parseImpl(mlir::AsmParser &parser, llvm::StringRef TypeName) {
   ::mlir::FailureOr<FieldsVectorType> Fields(FieldsVectorType{});
 
-  if (parser.parseOptionalRSquare().failed()) {
+  const auto ParseFieldsRSquare = [&]() -> bool {
```
Can you be more verbose in the commit message about what's being fixed (commit "Fix Clift type parsing ")?

vasama-work replied:
I added this explanation:
* Allow empty parameter list for FunctionAttr.
* Allow empty field list for struct but not union.

aleclearmind commented on include/revng-c/Tools/clift-opt/Model.h:
```diff
@@ -0,0 +1,9 @@
+
+#include "revng/Model/Binary.h"
+
+extern TupleTree<model::Binary> Model;
```
I guess this can become a `static` global in the `.cpp` file?

vasama-work replied:
I got rid of this and added a pass specific option. This was very ugly anyway. This model was filled from the command line by `clift-opt` `main` and used by the import model type passes.

aleclearmind commented on lib/mlir/Dialect/Clift/IR/CliftTypes.cpp:
```diff
@@ -307,3 +317,148 @@ StructType::replaceImmediateSubElements(llvm::ArrayRef<mlir::Attribute>
   return {};
 }
+
+/****************************** ScalarTupleType *******************************/
```
Avoid C-style comments

vasama-work replied:
I changed them to `//*** ScalarTupleType ***`. Let me know if you prefer some other style.

aleclearmind replied:
I usually do:
```c++
//
// ScalarTupleType
//
```
But we have no strict rule.

Alternatively you could maybe use doxygen grouping comments.
Alternatively, the fact that you have such a large portion of the file grouped together might hint at splitting those functions in another `.cpp` file.

vasama-work replied:
I'll leave it for now if that's okay. We can look it again in my next Clift cleanup PR.

aleclearmind commented on include/revng-c/mlir/Dialect/Clift/IR/CliftTypes.h:
```diff
@@ -15,8 +15,76 @@
+namespace mlir::clift {
+
+struct ScalarTupleTypeStorage;
+class ScalarTupleType : public Type::TypeBase<ScalarTupleType,
```
Add docs discussing why we need this.

aleclearmind commented on lib/mlir/Dialect/Clift/IR/CMakeLists.txt:
```diff
@@ -21,3 +21,9 @@ add_mlir_dialect_library(
   MLIRIR
   PRIVATE
   revng::revngModel)
+
```
Please add a TODO around all the usage of `add_mlir_dialect_library` to switch to `revng_add_library`.

`revng_add_library` does things that are important and we're bypassing it.

vasama-work replied:
I've added this in the internal Clift to-do document. Is that okay, or do you also want comments in the CMake files?

aleclearmind replied:
Add a comment in the CMake file too. Comments show that we now that the code we're looking at has a known limitation.

aleclearmind commented on tests/unit/CliftType.cpp:
```diff
@@ -0,0 +1,48 @@
+#include "revng-c/mlir/Dialect/Clift/IR/Clift.h"
+#include "revng-c/mlir/Dialect/Clift/Utils/ImportModel.h"
+
+// Don't reorder this
+#include "revng/tests/unit/ModelType.inc"
```
Usually this ends up at the end of the file. I guess you're forward declaring the static functions in there.
This can also work.

vasama-work replied:
I moved it to the end of the file. That works too.

aleclearmind commented on tests/unit/CliftType.cpp:
```diff
@@ -0,0 +1,48 @@
+#include "revng/tests/unit/ModelType.inc"
+
+template<typename CallableType>
+static auto withContext(CallableType Callable) {
```
Does `static` make sense on a template function? Aren't they implicitly `inline`?

aleclearmind commented on tests/unit/mlir_lit_tests/model-import/EnumType.yml:
```diff
@@ -0,0 +1,148 @@
+    # CHECK:   name = "my_enum2"
+    # CHECK:   underlying_type = !int64_t
+    # CHECK:   fields = [
+    # CHECK:     <
+    # CHECK:       raw_value = 0
+    # CHECK:     >
+    # CHECK:     <
+    # CHECK:       raw_value = 1
+    # CHECK:     >
+    # CHECK:     <
+    # CHECK:       raw_value = 9223372036854775807
+    # CHECK:     >
+    # CHECK:     <
+    # CHECK:       raw_value = 9223372036854775808
+    # CHECK:     >
+    # CHECK:     <
+    # CHECK:       raw_value = 18446744073709551615
+    # CHECK:     >
```
Lots of newlines, I wonder if this was intentional

vasama-work replied:
We do want to ignore other things such as enumerator name between the angle brackets. That's why I used multiple lines. Perhaps there is a way to do this in a single line using FileCheck, but I'm not sure.

aleclearmind replied:
Can you paste the full output so I can get a better idea?

vasama-work replied:
```CPP
!int32_t = !clift.primitive<SignedKind 4>
!int64_t = !clift.primitive<SignedKind 8>
!uint32_t = !clift.primitive<UnsignedKind 4>
!uint64_t = !clift.primitive<UnsignedKind 8>
#my_enum1_ = #clift.enum<id = 1104, name = "my_enum1", underlying_type = !int32_t, fields = [<name = "_enum_entry_my_enum1_0", raw_value = 0>, <name = "_enum_entry_my_enum1_1", raw_value = 1>, <name = "_enum_entry_my_enum1_2147483647", raw_value = 2147483647>, <name = "_enum_entry_my_enum1_18446744071562067968", raw_value = 18446744071562067968>, <name = "_enum_entry_my_enum1_18446744073709551615", raw_value = 18446744073709551615>]>
#my_enum2_ = #clift.enum<id = 1108, name = "my_enum2", underlying_type = !int64_t, fields = [<name = "_enum_entry_my_enum2_0", raw_value = 0>, <name = "_enum_entry_my_enum2_1", raw_value = 1>, <name = "_enum_entry_my_enum2_9223372036854775807", raw_value = 9223372036854775807>, <name = "_enum_entry_my_enum2_9223372036854775808", raw_value = 9223372036854775808>, <name = "_enum_entry_my_enum2_18446744073709551615", raw_value = 18446744073709551615>]>
#my_enum3_ = #clift.enum<id = 1204, name = "my_enum3", underlying_type = !uint32_t, fields = [<name = "_enum_entry_my_enum3_0", raw_value = 0>, <name = "_enum_entry_my_enum3_1", raw_value = 1>, <name = "_enum_entry_my_enum3_4294967295", raw_value = 4294967295>]>
#my_enum4_ = #clift.enum<id = 1208, name = "my_enum4", underlying_type = !uint64_t, fields = [<name = "_enum_entry_my_enum4_0", raw_value = 0>, <name = "_enum_entry_my_enum4_1", raw_value = 1>, <name = "_enum_entry_my_enum4_18446744073709551615", raw_value = 18446744073709551615>]>
!my_enum1_1 = !clift.defined<#my_enum1_>
!my_enum2_1 = !clift.defined<#my_enum2_>
!my_enum3_1 = !clift.defined<#my_enum3_>
!my_enum4_1 = !clift.defined<#my_enum4_>
module {
  %0 = clift.undef !int32_t
  %1 = clift.undef !int64_t
  %2 = clift.undef !uint32_t
  %3 = clift.undef !uint64_t
  %4 = clift.undef !my_enum1_1
  %5 = clift.undef !my_enum2_1
  %6 = clift.undef !my_enum3_1
  %7 = clift.undef !my_enum4_1
}
```

This is what is being matched: `<name = "_enum_entry_my_enum1_0", raw_value = 0>`

aleclearmind replied:
Aaaah, so `CHECK` does not necessarily imply a new line, didn't know that.

The fact that [`CHECK-SAME`](https://llvm.org/docs/CommandGuide/FileCheck.html#the-check-same-directive) exists made me think `CHECK` implied a new line 🤔 

Anyway, it'd be nice to have a more readable output for types where the type definition is split on multiple lines, but this works for now.

If you want to reduce the number of `CHECK` you can use [regex](https://llvm.org/docs/CommandGuide/FileCheck.html#filecheck-regex-matching-syntax) and match multiple things in a single shot.

vasama-work replied:
Yeah. I figure this is more readable than regexes though, even if it gets split up on multiple lines. And this will be less fragile if we ever add new parameters to enumerators.

aleclearmind replied:
Maybe add a comment that, even though the checks are split on multiple lines, we're testing a single line.

Other than that, :+1:.

aleclearmind commented on include/revng-c/Pipes/Kinds.h:
```diff
@@ -56,8 +56,8 @@ inline TypeKind ModelTypeDefinition("model-type-definition",
 
-inline pipeline::SingleElementKind
-  MLIRLLVMModule("mlir-llvm-module", Binary, ranks::Binary, {}, {});
+inline FunctionKind
+  MLIRLLVMFunctionKind("mlir-module", ranks::Function, {}, {});
```
Nice nice

aleclearmind commented on include/revng-c/mlir/Dialect/Clift/Utils/Helpers.h:
Can you please backlog (i.e., put in your TODO list for *after* this PR) to check if we LLVM infra requires this directory structure or if we can collapse `include/revng-c/mlir/Dialect/Clift/` to `include/Clift`?

vasama-work replied:
Done.

aleclearmind commented on include/revng-c/mlir/Dialect/Clift/Utils/Helpers.h:
```diff
@@ -0,0 +1,22 @@
+
+namespace mlir::clift {
+
+inline std::optional<MetaAddress> getMetaAddress(mlir::FunctionOpInterface F) {
```
Usually we don't use `std::optional<MetaAddress>`, `MetaAddress` already embeds an invalid encoding (i.e., `MetaAddress::invalid()`. Unless you have a really good reason, use `MetaAddress::invalid` instead of `std::nullopt`.

vasama-work replied:
Done.

vasama-work committed:
Add TODO on add_mlir_dialect_library

aleclearmind commented on include/revng-c/mlir/Pipes/MLIRContainer.h:
```diff
@@ -0,0 +1,65 @@
+
+  mlir::MLIRContext *getContext() { return Context.get(); }
+  mlir::ModuleOp getModule() { return *Module; }
+  void setModule(mlir::OwningOpRef<mlir::ModuleOp> &&NewModule);
```
I'm not super stoked about this method. What do you think about creating the module and the context in the constructor, and having users populated that one instead of creating their own and then `setModule` it?

vasama-work replied:
Usually that would be fine, but the import from LLVM returns a new module, which is where this function is needed.

aleclearmind replied:
Ah OK, so it's due to a "limitation" of some API we rely on. Gotcha.

aleclearmind commented on include/revng-c/mlir/Pipes/MLIRContainer.h:
```diff
@@ -0,0 +1,65 @@
+                         const pipeline::Target &Target) const override;
+
+  static std::vector<pipeline::Kind *> possibleKinds() {
+    return { &kinds::MLIRLLVMFunctionKind };
```
The name of the kind still embed LLVM, let's drop it.

aleclearmind commented on include/revng-c/mlir/Pipes/MLIRContainer.h:
```diff
@@ -0,0 +1,65 @@
+  }
+
+private:
+  void setModuleInternal(mlir::OwningOpRef<mlir::ModuleOp> &&NewModule);
```
This method is not defined/used anywhere.

vasama-work replied:
Good catch.

aleclearmind commented on lib/mlir/Pipes/MLIRContainer.cpp:
```diff
@@ -0,0 +1,354 @@
+static void visit(ModuleOp Module, Visitor visitor) {
+  auto &OpList = getModuleOperations(Module);
+
+  auto Begin = OpList.begin();
+  const auto End = OpList.end();
+
+  while (Begin != End) {
+    using RequestedOpT = decltype(visitHelper(&Visitor::operator()));
```
Why not a range-`for`?

vasama-work replied:
The iterator is advanced in the middle of the loop before calling the visitor. This is so that the visitor is able to safely erase the operation.

aleclearmind replied:
Uhm, OK, just make sure that the underlying data structure you're iterating actually offers the kind of iterator invalidation guarantees you expect.

vasama-work replied:
`OpList` is of type `llvm::iplist<mlir::Operation>` (see [here](https://mlir.llvm.org/doxygen/classmlir_1_1Block.html#af52a3ca2afbb4a1c29092f3888e78bb8)). As far as I can tell, no documentation for `iplist` explicitly states any iterator invalidation guarantees, but it's expected that erasing an element from a linked list does not invalidate iterators to other elements.

vasama-work replied:
I added explicit typing for `OpList` and an explanation of what's happening. See [here](https://github.com/revng/revng-c/pull/1/files#diff-24dded21127406a4fb4c2ef29c414dd57a94b946de77b22cf143ababf124cc4dR59).

aleclearmind replied:
OK, I just want to make sure you double checked that.

aleclearmind commented on lib/mlir/Pipes/MLIRContainer.cpp:
```diff
@@ -0,0 +1,354 @@
+  revng_assert(Operations.size() == 1);
+  auto DestinationModule = mlir::cast<ModuleOp>(Operations.front());
+
+#ifndef NDEBUG
+  const size_t SrcOpsCount = getModuleOperations(SourceModule).size();
+  const size_t NewOpsCount = getModuleOperations(DestinationModule).size();
+  revng_assert(NewOpsCount == SrcOpsCount);
+#endif
```
Nice. Sometimes, for expensive checks, we wrap the checks in

```c++
if (VerifyLog.isEnabled()) {
  // ...
}
```
Then you can enable them passing `--debug-log=verify`, which is used in testing and is useful if you get weird crashes. In future the flag will become something like `--extra-checks`, for now we abuse loggers.

aleclearmind commented on lib/mlir/Pipes/MLIRContainer.cpp:
```diff
@@ -0,0 +1,354 @@
+}
+
+// Remove any unused non-target symbols.
+static void pruneUnusedSymbols(ModuleOp Module) {
```
Maybe [-symbol-dce](https://mlir.llvm.org/docs/Passes/#-symbol-dce) is relevant here?

vasama-work replied:
Hard to say off the cuff... It looks like the non-target functions would at least have to be marked private at some point.

aleclearmind replied:
OK, maybe add a TODO comment about investigating this

aleclearmind commented on lib/mlir/Pipes/MLIRContainer.cpp:
```diff
@@ -0,0 +1,354 @@
+pipeline::TargetsList MLIRContainer::enumerate() const {
+  pipeline::TargetsList::List List;
+
+  visit(Module.get(), [&](FunctionOpInterface F) {
```
Is this visit recursive? I'd be sad if we were to got through all the operation just to enumerate functions.

vasama-work replied:
No. There's a comment to that effect on `visit` itself.

pfez commented on lib/mlir/Dialect/Clift/IR/CliftAttributes.cpp:
```diff
@@ -307,59 +308,62 @@ mlir::Attribute mlir::clift::StructType::parse(AsmParser &parser) {
+    return true;
+  }
+
+  return true;
```
:thinking: Perhaps this should `return false` also for `void`?

Actually, you're using this for verification of structs and unions. I believe that for that goal we can have something more appropriate that also returns `false` for function types, and in general for all types whose size is either unknown (like you do here for structs and unions forward declarations) or zero (like `void` and function types).

What do you think?

vasama-work replied:
Have a look at [this commit](https://github.com/revng/revng-c/pull/20/commits/565e42c5268695de4ba2db4b50dade42bbdbf56c) in my next PR.

pfez replied:
I checked, there are a few things that make sense, and some that are wrong. I left comments on the commit.

In any case pick that commit and make it part of this PR please, so that we merge things that are correct at the first shot.

And then we can fix the few issues that remain, that I've stressed in my comments there.

pfez replied:
Ok I'm closing this. Just pick the commit you linked above and bring it in this PR.

pfez commented on lib/mlir/Dialect/Clift/IR/CliftTypes.cpp:
```diff
@@ -141,8 +141,12 @@ PointerType::verify(::llvm::function_ref<::mlir::InFlightDiagnostic()>
                     BoolAttr IsConst) {
-  if (pointer_size == 0) {
-    return emitError() << "pointer type cannot have size zero";
+  switch (pointer_size) {
+  case 4:
+  case 8:
+    break;
+  default:
+    return emitError() << "invalid pointer size: " << pointer_size;
```
Eh sad. But I guess there's nothing better to do if we don't want the model to be involved. @aleclearmind any better idea?

pfez replied:
@aleclearmind ping!

At the moment I'd just leave it like this.
If anything we can double check this with assertions in our backend if we want extra safety.
But I don't want to pull too much model stuff in here.

pfez commented on lib/mlir/Dialect/Clift/IR/CliftAttributes.cpp:
```diff
@@ -175,6 +175,9 @@ static bool isCompleteType(const mlir::Type Type) {
 
+  if (auto T = mlir::dyn_cast<mlir::clift::ScalarTupleType>(Type))
+    return T.isComplete();
+
```
Can a `ScalarTupleType` ever be incomplete? I don't think so. In order for it be incomplete we'd need to have a scalar type that is itself incomplete, which is not possible, isn't it?

vasama-work replied:
Yes. It's needed for recursive types, just like with structs. Scalar tuples can be recursive when they contain a pointer to the function type that returns the tuple.

pfez replied:
Fair enough.

pfez commented on lib/mlir/Dialect/Clift/IR/CliftAttributes.cpp:
```diff
@@ -419,3 +422,15 @@ uint64_t mlir::clift::UnionType::getByteSize() const {
+mlir::clift::ScalarTupleElementAttr::verify(const EmitErrorType EmitError,
+                                            mlir::Type Type,
+                                            const llvm::StringRef Name) {
+  if (not mlir::isa<mlir::clift::ValueType>(Type))
+    return EmitError() << "Scalar tuple element type must be a value type";
+
```
We should also verify that it's a scalar.

Scalars are non-void primitives, pointers, and their typedefs.

vasama-work replied:
See the discussion regarding `isCompleteType` above.

pfez replied:
You mean [this](https://github.com/revng/revng-c/pull/1#discussion_r1611226825) ?

vasama-work replied:
That's right. I already have a lot of upcoming changes to improve type verification in the next PR. If you really want me to, I can move it to this one too.

pfez replied:
Ok this is fine now!

pfez commented on include/revng-c/mlir/Dialect/Clift/IR/CliftAttributes.td:
```diff
@@ -130,4 +130,17 @@ def Clift_FunctionAttr : Clift_Attr<"Function", "function", [DeclareAttrInterfac
+  let genVerifyDecl = 1;
+
+  let assemblyFormat = "`<`struct($name, $type)`>`";
+}
```
Ok cool, we have `ScalarTupleElementAttr` defined in tablegen. But in principle it should be possible to have `ScalarTupleType` defined in tablegen too! Pretty much like we do e.g. for enums and typedefs.

We're forced to go the full C++ path for structs and unions because of the bad support of recursive types, but scalar tuple types are not fully recursive, so we should stick with tablegen.

Unless there's some other reason I'm missing?

pfez replied:
I'll wait reviewing the rest of this commit until we've sorted out this issue.

vasama-work replied:
Scalar tuples can be recursive too.

pfez replied:
Right

pfez commented on lib/mlir/Dialect/Clift/IR/CliftAttributes.cpp:
```diff
@@ -344,14 +348,19 @@ static bool isCompleteType(const mlir::Type Type) {
+                                                    uint64_t Size) {
+  if (Size == 0)
+    return EmitError() << "struct type cannot have a size of zero";
+
```
Can't this just call the following with empty `Fields`?

pfez commented on lib/mlir/Dialect/Clift/IR/CliftParser.h:
```diff
@@ -0,0 +1,201 @@
+
+namespace mlir::clift {
+
+class AsmRecursionGuardBase {
+  using Map = std::map<uint64_t, void *>;
+
+  std::pair<Map::iterator, bool> Insert;
+
+public:
+  explicit AsmRecursionGuardBase(const uint64_t ID);
+
+  AsmRecursionGuardBase(const AsmRecursionGuardBase &) = delete;
+  AsmRecursionGuardBase &operator=(const AsmRecursionGuardBase &) = delete;
+
+  ~AsmRecursionGuardBase();
+
+protected:
+  void *get() const { return Insert.first->second; }
+
+  void set(void *const Value) {
+    revng_assert(Insert.second);
+    revng_assert(Insert.first->second == nullptr);
+    Insert.first->second = Value;
+  }
+};
+
+template<typename ObjectT>
+class AsmRecursionGuard : AsmRecursionGuardBase {
+public:
+  using AsmRecursionGuardBase::AsmRecursionGuardBase;
+
+  [[nodiscard]] ObjectT get() const {
+    return ObjectT(static_cast<
+                   typename ObjectT::ImplType *>(AsmRecursionGuardBase::get()));
+  }
+
+  void set(ObjectT Value) { AsmRecursionGuardBase::set(Value.getImpl()); }
+};
```
While I understand the desire to make a better Guard, I think the result is more confusing than the previous version with `getVisitedType` and `onTypeVisited`.

Now we have this wrappers with more layers of abstraction, with an API `get`/`set` that has non-informative naming.
Honestly I find harder to convince myself that the current version of the client code is bug-free than with the previous version. It feels a little bit over-engineered for solving a specific local problem.

So, pick one of the following:
1. revert to `getVisitedType` and `onTypeVisited`
2. improve and document the new guard
  - why `Asm` in the name?
  - what is the general problem it tries to solve? with the previous names it was kinda easier to guess, now it isn't
  - what are the contracts of constructors/destructors?
  - what are the contracts of `get`/`set`
  - do we really need the `Base` version? on the one hand you add a template parameter, but then you use `void *` under the hood :thinking:
  
Honestly I'd just stick with 1.

vasama-work replied:
> why Asm in the name?

Because it relates to the parsing and printing of MLIR assembly. Same as `mlir::AsmPrinter` etc.

> what is the general problem it tries to solve?

The main point is to use RAII for safety. With this, the user cannot forget to erase the key from the map, even in the case of early returns.

> do we really need the `Base` version?

That allows me to define the thread local out-of-line, which is something I do as a rule. The codegen for inline thread locals is atrocious.

I can add comments explaining these things, but I much prefer 2 over 1 just for the RAII. As an alternative, if we have a scope guard utility, I could use that as well. But I really want to do the clean up in a destructor.

vasama-work replied:
Apparently we don't have a generic scope guard, so if we want to go that route, one would have to be added. That would go in `revng` I suppose.

vasama-work replied:
Oh, I totally forgot that LLVM has one which was already being used here.

pfez replied:
> Oh, I totally forgot that LLVM has one which was already being used here.

Great!

vasama-work replied:
Alright, I got rid of those classes. See [here](https://github.com/revng/revng-c/pull/1/commits/ba1cd69cb5bb7d14b54f5cbc55e2e0d86572a289) for a diff. I will squash that commit into the old one once you're happy with it.

vasama-work replied:
It's the `Review changes (to be squashed)` commit. The link gets outdated when I make changes.

pfez replied:
Ah lovely!

pfez commented on lib/mlir/Dialect/Clift/Utils/ImportModel.cpp:
```diff
@@ -0,0 +1,589 @@
+
+    ~RecursiveDefinitionGuard() {
+      if (Self != nullptr) {
+        Self->DefinitionGuardSet.erase(ID);
```
Shouldn't this assert that it has erased something?

pfez commented on lib/mlir/Dialect/Clift/Utils/ImportModel.cpp:
```diff
@@ -0,0 +1,589 @@
+
+  mlir::BoolAttr getFalse() { return getBool(false); }
+
+  template<typename T, typename... ArgTypes>
+  T make(const ArgTypes &...Args) {
+    if (failed(T::verify(EmitError, Args...)))
+      return {};
+    return T::get(Context, Args...);
+  }
```
Isn't calling this equivalent to just calling `T::getChecked(Context, Args...)`? If so this is basically a just a helper for conciseness right?

vasama-work replied:
Yes. It used to be a little different, but I got rid of that and now it might be equivalent to just calling `getChecked`, except that `getChecked` here doesn't compile for all types. I didn't investigate it deeply enough to be sure on the root cause of that issue.

pfez commented on lib/mlir/Dialect/Clift/Utils/ImportModel.cpp:
```diff
@@ -0,0 +1,589 @@
+      rc_return nullptr;
+    }
+
+    // Qualifier::isConst verifies itself, which interferes with unit testing.
+    const auto IsConst = [](const model::Qualifier &Q) {
+      return Q.Kind() == model::QualifierKind::Const;
+    };
+
+    // Qualifier::isPointer verifies itself, which interferes with unit testing.
+    const auto IsPointer = [](const model::Qualifier &Q) {
+      return Q.Kind() == model::QualifierKind::Pointer;
+    };
```
What do you mean they interfere with unit testing?

vasama-work replied:
They contain assertions verifying the type and some of the unit tests are testing for import failure. Using those functions would cause the assertions to fail and in turn cause the unit test to fail.

pfez replied:
Aaah no ok, I finally got what you meant here! It's ok then!

pfez commented on lib/mlir/Dialect/Clift/Utils/ImportModel.cpp:
```diff
@@ -0,0 +1,589 @@
+
+    auto Qualifiers = llvm::ArrayRef(ModelType.Qualifiers());
+
+    const auto takeQualifier = [&]() -> const model::Qualifier & {
+      const model::Qualifier &Qualifier = Qualifiers.back();
+      Qualifiers = Qualifiers.slice(0, Qualifiers.size() - 1);
+      return Qualifier;
+    };
+
+    const auto takeConst = [&]() -> bool {
+      if (not Qualifiers.empty() and IsConst(Qualifiers.back())) {
+        Qualifiers = Qualifiers.slice(0, Qualifiers.size() - 1);
+        return true;
+      }
+      return false;
+    };
```
Declare these ALAP, and only capture `Qualifiers` please.

pfez commented on lib/mlir/Dialect/Clift/Utils/ImportModel.cpp:
```diff
@@ -0,0 +1,589 @@
+    if (RequireComplete and llvm::any_of(Qualifiers, IsPointer))
+      RequireComplete = false;
+
+    clift::ValueType ResultType = rc_recur
+      getUnwrappedValueType(*ModelType.UnqualifiedType().get(),
+                            RequireComplete,
+                            takeConst());
```
Ah so, the "last" qualifier is pushed inside here and managed by `getUnwrappedValueType`.

Hopefully this will be simpler and less awkward with ivan's new patchset on model types.

vasama-work replied:
Yes, I reckon this will become much simpler.

pfez commented on include/revng-c/mlir/Dialect/Clift/Transforms/Passes.h:
```diff
@@ -0,0 +1,19 @@
+
+namespace mlir::clift {
+
+#define GEN_PASS_DECL
+#include "revng-c/mlir/Dialect/Clift/Transforms/Passes.h.inc"
+
+std::unique_ptr<OperationPass<mlir::ModuleOp>> createImportModelTypesPass();
+
+#define GEN_PASS_REGISTRATION
+#include "revng-c/mlir/Dialect/Clift/Transforms/Passes.h.inc"
```
So this is necessary to define an `OperationPass` and register it. And it's including an `.inc` file that is generated by tablegen by de declarative stuff in `include/revng-c/mlir/Dialect/Clift/Transforms/Passes.td`. Is that right?

I'm taking this as an opportunity to understand all the mlir tricks and make sure I remember them.

vasama-work replied:
Right. It's just including the tablegen generated stuff.

pfez commented on lib/mlir/Dialect/Clift/Transforms/ImportModelTypes.cpp:
```diff
@@ -0,0 +1,55 @@
+#include "revng-c/mlir/Dialect/Clift/Transforms/Passes.h"
+#include "revng-c/mlir/Dialect/Clift/Utils/ImportModel.h"
+
+namespace mlir {
+namespace clift {
+#define GEN_PASS_DEF_CLIFTIMPORTMODELTYPES
+#include "revng-c/mlir/Dialect/Clift/Transforms/Passes.h.inc"
+} // namespace clift
+} // namespace mlir
```
And this is another piece of the magic. How does this part work?

vasama-work replied:
More tablegen generated stuff but for the implementation this time. Also unlike the header stuff, the implementation stuff only provides per pass opt-in macros. The header stuff also has these but also has macros to include all. (`GEN_PASS_DECL`, `GEN_PASS_REGISTRATION`). In particular on the implementation side it defines the `CliftImportModelTypesBase` base class from which the pass implementation is derived.

aleclearmind commented:
@pfez this is ready for me, please mark it as ready and inform me once it's ready for you too.

pfez commented on lib/mlir/Dialect/Clift/Utils/ImportModel.cpp:
```diff
@@ -0,0 +1,585 @@
+
+    // If the set of qualifiers contains any pointers,
+    // the base type does not need to be complete.
+    if (RequireComplete and llvm::any_of(Qualifiers, IsPointer))
```
Now that I think about it, we have `model::Qualifier::isPointer` instead of `IsPointer`.

We also have `mode::Qualifier::isConst` instead of `IsConst` below.

pfez replied:
Forget about this. I finally understood what you meant in you other comment.

pfez commented:
@aleclearmind this is ready for me! I'm adding the "ready" label!

There's only a very minor issue but there's already a commit in https://github.com/revng/revng-c/pull/20 that fixes, and I expect to merge that soon too, so whenever you want this can already be merged.

vasama-work committed:
Move CliftStorage.h to lib/

vasama-work committed:
Add ScalarTupleType for RawFunctionType return

vasama-work committed:
Refactor StructTypeStorage and UnionTypeStorage

vasama-work committed:
Naming and reorganization for consistency

vasama-work committed:
Fix Clift type parsing

* Allow empty parameter list for FunctionAttr.
* Allow empty field list for struct but not union.

pfez commented:
Merged in 15227190d9faf457c146c59bf7c3ff1664655b71

Thanks!
